### PR TITLE
[AppRewrite] Part 6: Loaders in the `telestion-services` module

### PR DESCRIPTION
**Pull request by fussel178**
### Stacked Pull Request

This is a stacked pull request. Please wait on pending pull requests that reference on this branch before merging this branch.
Take a look at https://www.michaelagreiler.com/stacked-pull-requests/ for more information.

### Summary 

This PR adds ready-to-use Loaders to the `telestion-services` module for the most common use cases.

#### The Debug Loader

The Debug Loader implements every possible event hook and logs a debug message.
This Loader is only useful to debug and test different Loader Launcher implementations.
It doesn't accept any configuration.

You can use it via its classname:

`conf/telestion.json`
```json
{
  "loaders": [
    "de.wuespace.telestion.services.loader.DebugLoader"
  ]
}
```

#### The Abstract Verticle Loader

> Note: This Loader is not usable in the `telestion.json`. Please use the Verticle Loader or the Config Loader instead.

This Loader extends the functionality of the TelestionLoader and provides methods to deploy verticles from a Verticle Configuration.
The Abstract Verticle Loader accepts one configuration property `verticlesProperty` which specifies the key the Loader should use to access the Verticle Configurations from the given configuration.
It provides the `onConfiguration` method that extending Loaders should use to deploy Verticle Configurations from a main configuration.

An example could look like:

```java
public class MyVerticleLoader extends AbstractVerticleLoader<MyVerticleLoader.Configuration> {
	public record Configuration(
			@JsonProperty String verticlesProperty
	) implements VerticleLoaderConfiguration {
	}

	@Override
	public void onAfterVertxStartup(Promise<Void> startPromise) throws Exception {
		// load the main configuration from a specific place
		var mainConfiguration = load(/* ... */);
		onConfiguration(mainConfiguration).onComplete(startPromise);
	}
}
```

#### The Verticle Loader

The Verticle Loader reads the Verticle Configurations from the main configuration of the associated Launcher.
It has the same configuration as the Abstract Verticle Loader and use it as basis.

You can import the Loader via its classname:

`conf/telestion.json`
```json
{
  "loaders": [
    "de.wuespace.telestion.services.loader.VerticleLoader"
  ],
  "verticles": [
    {
      "name": "Message Logger",
      "verticle": "de.wuespace.telestion.services.monitoring.MessageLogger",
      "magnitude": 1,
      "config": {}
    }
  ]
}
```

#### The Config Loader

> Note: This Loader covers a rather advanced use case. If you want to load Verticle Configurations from the main configuration please see the Verticle Loader.

The Config Loader uses the [Vert.x Config Retriever](https://vertx.io/docs/vertx-config/java/) to load and merge the configuration from multiple stores (like the good old `conf/config.json`).

This Loader accepts beside the `verticlesProperty` a configuration object which maps to the `ConfigRetrieverOptions`.

You can use it via its classname and additional configuration:

`conf/telestion.json`
```json
{
  "loaders": [
    {
      "className": "de.telestion.wuespace.services.loader.ConfigLoader",
      "config": {
        "includeDefaultStores": true,
        "stores": []
      }
    }
  ]
}
```

`conf/config.json`
```json
{
  "verticles": [
    {
      "name": "Message Logger",
      "verticle": "de.wuespace.telestion.services.monitoring.MessageLogger",
      "magnitude": 1,
      "config": {}
    }
  ]
}
```

### Sample project

You can find a sample project that uses the new application structure here:

https://github.com/wuespace/telestion-project-playground/tree/fussel178/feat/application-rewrite/application

### Next steps

This PR belongs to a series of changes which updates the application structure in Telestion.

Read previous: https://github.com/wuespace/telestion-core/pull/504

### CLA

- [x] I have signed the individual contributor's license agreement **and sent** it to the board of the WüSpace e. V. organization.

### CODE REVIEW

File: modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/verticle/AbstractVerticleLoader.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+public abstract class AbstractVerticleLoader<T extends VerticleLoaderConfiguration> extends TelestionLoader<T> {
+
+	/**
+	 * The default property name which points to the {@link VerticleConfiguration VerticleConfigurations}
+	 * in the loader configuration.
+	 */
+	public static final String DEFAULT_VERTICLES_PROPERTY = "verticles";
+
+	/**
+	 * Gets called, when the loader receive verticles configuration.
```
Comment (cb0s) [inline modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/verticle/AbstractVerticleLoader.java:33]:
```suggestion
	 * Gets called, when the loader receives verticle configurations.
```

File: modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/verticle/AbstractVerticleLoader.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	 *
+	 * @param configuration the retrieved configuration
+	 * @return a future which resolves when all verticles are deployed
```
Comment (cb0s) [inline modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/verticle/AbstractVerticleLoader.java:36]: Add Link to Future, pls :)

(multiple times)

File: modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/verticle/VerticleConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import de.wuespace.telestion.api.message.JsonMessage;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * @author Jan von Pichowski (@jvpichowski), Ludwig Richter (@fussel178)
+ */
+public record VerticleConfiguration(
```
[the remark this answers is not in the document]
Comment (cb0s) [reply modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/verticle/VerticleConfiguration.java:11]: And please add a description about the different properties of this record :)

File: modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/verticle/VerticleLoaderConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package de.wuespace.telestion.services.loader.verticle;
+
+import de.wuespace.telestion.application.loader.LoaderConfiguration;
+
+public interface VerticleLoaderConfiguration extends LoaderConfiguration {
```
Comment (cb0s) [inline modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/verticle/VerticleLoaderConfiguration.java:5]: Please let nobody make the song "We don't need no Java-Docs... 🎸" about our repository :D

**Review by cb0s — CHANGES_REQUESTED** [review]
I mostly need more java docs. Apart from that as already stated in the previous PRs of the stacked PR: VERY GOOOD! ✔️

File: modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/VerticleLoader.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	 * @param verticlesProperty the property name which points to the verticle configuration
+	 */
+	public record Configuration(
+			@JsonProperty String verticlesProperty
+	) implements VerticleLoaderConfiguration {
+	}
+
+	@Override
+	public void onAfterVertxStartup(Promise<Void> startPromise) throws Exception {
+		onConfiguration(launcher.getMainConfiguration()).onComplete(startPromise);
```
[the remark this answers is not in the document]
Comment (fussel178) [reply modules/telestion-services/src/main/java/de/wuespace/telestion/services/loader/VerticleLoader.java:54]: Done in [`ff4e091` (#505)](https://github.com/wuespace/telestion-core/pull/505/commits/ff4e091c2df60481b114d87da6bc968a6a2027b5)

**Comment by fussel178** [comment]
I'll close this pull request due to inactivity and re-create one pull request with a simpler implementation.
In this new pull request the [RFC: $ (dollar sign) operators in Application Configuration](https://github.com/wuespace/telestion/discussions/347) will also be addressed.

**Pull request closed without merging**